keps/sig-node/5359-pod-level-swap-control/README.md

```diff
@@ -0,0 +1,1118 @@
+
+The initial focus is to allow pods to explicitly disable swap, even if swap is enabled at the node level
+(via the `LimitedSwap` setting introduced in [KEP-2400](https://github.com/kubernetes/enhancements/issues/2400)).
+In the future, the pod-level swap control API field can be extended to allow more granular control over swap usage,
+or to provide hints to kubelet on how to configure swap for individual pods, depending on the configured "swapBehavior".
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Kubernetes has introduced node-level swap support (KEP-2400), currently Beta3 in 1.33.
+KEP-2400 allows nodes to utilize swap memory by configuring kubelet to use the `LimitedSwap` swap behavior.
+The `LimitedSwap` mode is currently purely automatic and implicit.
+Swap limits are automatically assigned to
+pods, [based on various factors](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#steps-to-calculate-swap-limit),
+while users cannot specifically configure swap for individual pods.
+For example, `LimitedSwap` restricts Guaranteed and high-priority pods from swap for performance reasons,
+and users cannot specifically exclude other critical pods from potential swap usage.
+
+While node-level swap can improve node stability and resource utilization in certain scenarios, it presents a challenge
+to some latency-sensitive applications.
+For these applications, performance degradation with any swap activity may be undesirable.
+Currently, if a node has swap enabled, application owners do not have a standard Kubernetes API mechanism to express
+their workload's intolerance to swap.
+
+This KEP addresses this gap by providing a pod-level swap control API.
+In the first phase, we'll focus on the ability to disable swap for all its containers irrespective of underlying node
+swap behavior.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* Introduce a new `swapPolicy` field in PodSpec to allow users to configure swap for an individual pod.
+* Initially, provide a way to disable swap under the `LimitedSwap` swap behavior.
+* Maintain backward compatibility: existing pods that run on swap-enabled nodes should behave as they do by default.
+* Provide a mechanism to alleviate concerns regarding "all-or-nothing" nature of node-level swap enablement, potentially
+  unblocking KEP-2400’s path to GA.
+* Open the door to future enhancements, both at the kubelet-level swap behaviors and the new pod-level swap control API
```

Comment:
so this covers the transition from 2400 to the ability to disable swap. Why we are not covering how we will transition to specifying the limit explicitly? We need at least an idea how pods with the limits specified will work with kubelet with limited swap. Or if they will be completely incompatible, than we need to explicitly say that LimitedSwap is a limited time feature and will be fully replaced with something else

Reply:
I think pod level swap can be another swap type for the pod. Pod API always wins, if kubelet is limited swap then we also add swap if the pod said "NoPreference", or if it's disabled in kubelet and NoPreference in api then no swap

Reply:
also: if node doesn't have swap enabled, then pod API swap pods fail at admission (maybe with NodeCapabilities feature to help scheduling hints)

Reply:
I suggested something here: https://docs.google.com/document/d/19muCmfBndxfr-yq1EmeI54yq-upYmZsWLsLvcm1i6J4/edit?disco=AAABljwWs3w In general it feels we may have a path forward